### PR TITLE
Refactor `QueryBackend` and `QueryBackendParams`

### PR DESCRIPTION
**Pull request by errfrom**
### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: src/Internal/Contract/Monad.purs [hunk 1]
```diff
@@ -60,11 +54,7 @@ import Ctl.Internal.QueryM
   , mkOgmiosWebSocketAff
   , underlyingWebSocket
   )
--- TODO Move/translate these types into Cardano
```
Comment (rynoV) [inline src/Internal/Contract/Monad.purs:63]: Does this TODO have a resolution or an issue?

File: src/Internal/Contract/Monad.purs [hunk 2]
```diff
@@ -197,7 +187,7 @@ mkContractEnv params = do
     b1 <- parallel do
       backend <- buildBackend logger params.backendParams
       -- Use the default backend to fetch ledger constants
```
Comment (rynoV) [inline src/Internal/Contract/Monad.purs:189]: Comment seems redundant

File: src/Internal/Contract/Monad.purs [hunk 3]
```diff
@@ -257,11 +248,11 @@ getLedgerConstants
        , systemStart :: Ogmios.SystemStart
        }
 getLedgerConstants logger = case _ of
-  CtlBackend { ogmios: { ws } } -> do
+  CtlBackend { ogmios: { ws } } _ -> do
     pparams <- getProtocolParametersAff ws logger
     systemStart <- getSystemStartAff ws logger
     pure { pparams, systemStart }
-  BlockfrostBackend _ -> undefined
+  BlockfrostBackend _ _ -> undefined
```
Comment (rynoV) [inline src/Internal/Contract/Monad.purs:255]: Will this `undefined` not be an issue? For example [here](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1314/files#diff-cd9604d9a86ee84c0def99f84ddea1aa930a7b511a2ada6493fdea9e61ced7cfR190)

File: src/Internal/Contract/Monad.purs [hunk 4]
```diff
@@ -279,18 +270,20 @@ walletNetworkCheck envNetworkId wallet = do
...
+stopContractEnv { backend } = liftEffect do
+  traverse_ stopCtlRuntime (getCtlBackend backend)
+  traverse_ stopBlockfrostRuntime (getBlockfrostBackend backend)
+  where
+  stopCtlRuntime :: CtlBackend -> Effect Unit
+  stopCtlRuntime { ogmios, odc } =
+    stopWebSocket odc.ws *> stopWebSocket ogmios.ws
+
+  stopBlockfrostRuntime :: BlockfrostBackend -> Effect Unit
+  stopBlockfrostRuntime = undefined
```
Comment (rynoV) [inline src/Internal/Contract/Monad.purs:283]: Why include this if it doesn't do anything?

**Review by rynoV — CHANGES_REQUESTED** [review]
Just a few things, but I agree this is a better way of doing it

File: src/Internal/Contract/Monad.purs (hunk 3 above)
Comment (Luis-omega) [reply src/Internal/Contract/Monad.purs:255]: I think it's okay as long as we don't export to the `BlockfrostBackend` to public interface yet. After all the point of #1119  is to  be able to work this in #1120

**Review by Luis-omega — APPROVED** [review]
I like this more, is less configurable but I don't think we have to worry about more backends in a while.

File: src/Internal/Contract/Monad.purs (hunk 3 above)
Comment (rynoV) [reply src/Internal/Contract/Monad.purs:255]: Ah I see, thanks for the info

File: src/Internal/Contract/Monad.purs (hunk 1 above)
Comment (errfrom) [reply src/Internal/Contract/Monad.purs:63]: Not yet, restored this TODO

File: src/Internal/Contract/Monad.purs (hunk 2 above)
Comment (errfrom) [reply src/Internal/Contract/Monad.purs:189]: But we still use the default backend, not the optional one

**Review by jy14898 — APPROVED** [review]
Yeah I prefer this design, it's always nicer when your invariants are directly encoded in the datatype

File: src/Internal/Contract/QueryBackend.purs [hunk 5]
```diff
@@ -136,21 +41,34 @@ type BlockfrostBackend =
...
   }
 
-data QueryBackend
-  = CtlBackend CtlBackend
-  | BlockfrostBackend BlockfrostBackend
+getCtlBackend :: QueryBackend -> Maybe CtlBackend
+getCtlBackend (CtlBackend backend _) = Just backend
+getCtlBackend (BlockfrostBackend _ backend) = backend
+
+getBlockfrostBackend :: QueryBackend -> Maybe BlockfrostBackend
```
Comment (uhbif19) [inline src/Internal/Contract/QueryBackend.purs:48]: Why do we need those? And `Maybe` parameter at all?

**Pull request merged**